api/src/model/request/play_asset_pack_fetch_request.gd

```diff
@@ -0,0 +1,120 @@
+# ##############################################################################
+#
+#	Copyright 2020 Google LLC
+#
+#	Licensed under the Apache License, Version 2.0 (the "License");
+#	you may not use this file except in compliance with the License.
+#	You may obtain a copy of the License at
+#
+#		https://www.apache.org/licenses/LICENSE-2.0
+#
+#	Unless required by applicable law or agreed to in writing, software
+#	distributed under the License is distributed on an "AS IS" BASIS,
+#	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#	See the License for the specific language governing permissions and
+#	limitations under the License.
+#
+# ##############################################################################
+#
+# Request object that handles asynchronous logic related to fetch_asset_pack().
+#
+# This object provides relevant getters so that it is possible to retrieve
+# the updated states from this object once the request completes.
+#
+# ##############################################################################
+class_name PlayAssetPackFetchRequest
+extends PlayAssetDeliveryRequest
+
+# -----------------------------------------------------------------------------
+# Emits request_completed(did_succeed, pack_name, result, exception) signal 
+# upon request succeeds/fails.
+# 	did_succeed : boolean indicating request succeeded/failed
+# 	pack_name: String, name of the requested asset pack
+# 	result : PlayAssetPackState object if request succeeded, otherwise null
+#	exception: PlayAssetPackException object if request failed, otherwise null
+#
+# Note: when calling fetch_asset_pack() on a non-existent pack_name, 
+# did_succeed will be false and both result and exception will be null.
+# 
+# Emits state_updated(pack_name, result) signal upon fetched pack's state 
+# updated. Only available if request completed success status.
+# 	pack_name: String, name of the requested asset pack
+# 	result: most up-to-date PlayAssetPackState object
+# -----------------------------------------------------------------------------
+signal request_completed(did_succeed, pack_name, result, exception)
+signal state_updated(pack_name, result)
+
+var _pack_name : String
+var _did_succeed : bool
+var _state : PlayAssetPackState
+var _error : PlayAssetPackException
+
+func _init(pack_name):
+	_pack_name = pack_name
+
+# -----------------------------------------------------------------------------
+# Release the reference of this Request object in PlayAssetPackManager 
+# singleton, then calls free() in superclass.
+# -----------------------------------------------------------------------------
+func free():
+	PlayAssetPackManager._remove_fetch_request_reference(self)
```

Comment:
In current design, developers need to manually call ```free()``` to fully release the references.

Reply:
Another way to solve this is automatically release the references in PlayAssetPackManager when reaching  {CANCELED, COMPLETED, FAILED, NOT_INSTALLED, UNKNOWN} state. The downside of this solution is that developers have no full control over this reference once download starts. Which solution do you think is better?

Reply:
Discussed offline: remove ```free()``` and releases reference automatically.